src/ngraph/runtime/gpu/gpu_emitter.cpp

```diff
@@ -1038,35 +1048,53 @@ void runtime::gpu::GPU_Emitter::emit_Reduce(EMIT_ARGS)
                     // Work around a compiler warning (*node inside typeid may have effects
                     // with shared pointers, which is fine here but clang doesn't like it.)
                     auto& fn = *op;
-                    auto f_ptr = reduce_map.find(type_index(typeid(fn)));
-                    if (f_ptr == reduce_map.end())
-                    {
-                        throw runtime_error("reduce with function " + fn.get_name() +
-                                            " is not implement yet.");
-                    }
-                    else if (op_count != 1)
+                    op_name = fn.get_name();
+                    reduce_op = op;
+                    if (op_count != 1)
                     {
                         throw runtime_error("reduce with more than one op is not implement yet.");
                     }
-                    else
-                    {
-                        reduce_tensor_op = f_ptr->second;
-                    }
                 }
-                std::vector<element::Type> dtypes{args[0].get_element_type(),
-                                                  out[0].get_element_type()};
-                auto& cudnn_emitter =
-                    external_function->get_primitive_emitter()->get_cudnn_emitter();
-                auto reduce_index =
-                    cudnn_emitter->build_reduce_forward(reduce_tensor_op,
-                                                        dtypes,
-                                                        args[0].get_shape(),
-                                                        reduce_op->get_reduction_axes(),
-                                                        CUDNNEmitter::ReductionMode::Reduce);
 
+                auto& fn = *reduce_op;
+                if (TI(fn) == TI(ngraph::op::Add))
```

Comment:
We really need a better way to test if a node is a particular type.

Reply:
We should down cast here `if (std::dynamic_pointer_cast<ngraph::op::Add>(fn))`

Reply:
maybe node should have a enum of unique name.

Reply:
Can we change these conditionals to use dynamic_pointer_casts? This is the proper way of using RTTI in this case. type_index works, but it is a bit more expensive